pkg/server/v3/delta_test.go

```diff
@@ -0,0 +1,393 @@
+			&cache.RawDeltaResponse{
+				SystemVersionInfo: "6",
+				Resources:         []types.Resource{runtime},
+				DeltaRequest:      &discovery.DeltaDiscoveryRequest{TypeUrl: rsrc.RuntimeType},
+			},
+		},
+		// Pass-through type (types without explicit handling)
+		opaqueType: {
+			&cache.RawDeltaResponse{
+				SystemVersionInfo: "7",
+				Resources:         []types.Resource{opaque},
+				DeltaRequest:      &discovery.DeltaDiscoveryRequest{TypeUrl: opaqueType},
+			},
+		},
+	}
+}
+
+func process(typ string, resp *mockDeltaStream, s server.Server) error {
+	var err error
+	switch typ {
+	case rsrc.EndpointType:
+		err = s.DeltaEndpoints(resp)
+	case rsrc.ClusterType:
+		err = s.DeltaClusters(resp)
+	case rsrc.RouteType:
+		err = s.DeltaRoutes(resp)
+	case rsrc.ListenerType:
+		err = s.DeltaListeners(resp)
+	case rsrc.SecretType:
+		err = s.DeltaSecrets(resp)
+	case rsrc.RuntimeType:
+		err = s.DeltaRuntime(resp)
+	case opaqueType:
+		err = s.DeltaAggregatedResources(resp)
+	}
+
+	return err
+}
+
+func TestDeltaResponseHandlersWildcard(t *testing.T) {
+	for _, typ := range testTypes {
+		t.Run(typ, func(t *testing.T) {
+			config := makeMockConfigWatcher()
+			config.deltaResponses = makeDeltaResponses()
+			s := server.NewServer(context.Background(), config, server.CallbackFuncs{})
+
+			resp := makeMockDeltaStream(t)
+			// This is a wildcard request since we don't specify a list of resource subscriptions
+			resp.recv <- &discovery.DeltaDiscoveryRequest{Node: node, TypeUrl: typ}
+
+			go func() {
+				err := process(typ, resp, s)
+				if err != nil {
+					t.Errorf("Delta() => got \"%v\", want no error", err)
+				}
+			}()
+
+			select {
+			case res := <-resp.sent:
+				close(resp.recv)
+
+				if config.deltaCounts[typ] != 1 {
+					t.Errorf("watch counts for typ: %s => got %v, want 1", typ, config.deltaCounts[typ])
+				}
+
+				if v := res.GetSystemVersionInfo(); v != "" {
+					t.Errorf("expected emtpy version on initial request, got %s", v)
+				}
+			case <-time.After(1 * time.Second):
+				t.Fatalf("got no response")
+			}
+		})
+	}
+}
+
+func TestDeltaResponseHandlers(t *testing.T) {
+	for _, typ := range testTypes {
+		t.Run(typ, func(t *testing.T) {
+			config := makeMockConfigWatcher()
+			config.deltaResponses = makeDeltaResponses()
+			s := server.NewServer(context.Background(), config, server.CallbackFuncs{})
+
+			resp := makeMockDeltaStream(t)
+			// This is a wildcard request since we don't specify a list of resource subscriptions
+			res, err := config.deltaResponses[typ][0].GetDeltaDiscoveryResponse()
+			if err != nil {
+				t.Error(err)
+			}
+			// We only subscribe to one resource to see if we get the appropriate number of responses back
+			resp.recv <- &discovery.DeltaDiscoveryRequest{Node: node, TypeUrl: typ, ResourceNamesSubscribe: []string{res.Resources[0].Name}}
+
+			go func() {
+				err := process(typ, resp, s)
+				if err != nil {
+					t.Errorf("Delta() => got \"%v\", want no error", err)
```

Comment:
This Delta() bit seems confusing, reword? Here and in other tests

Reply:
Might be too much for this PR but I find something like

```
err := process(typ, resp, s)
assert.NoError(t, err)
```

to be a lot more readable, and you don't have to think up a clever error string every time :)

Reply:
100% I use the assert testing in my other projects. I'll bring it into here it's much easier to read